pkg/syncer/spec/spec_process.go

```diff
@@ -171,43 +172,63 @@ func (c *Controller) process(ctx context.Context, gvr schema.GroupVersionResourc
 			err = c.downstreamClient.Resource(gvr).Delete(ctx, name, metav1.DeleteOptions{})
 		}
 		if err != nil && !apierrors.IsNotFound(err) {
-			return err
+			return nil, err
 		}
-		return nil
+		return nil, nil
 	}
 
 	// upsert downstream
 	upstreamObj, ok := obj.(*unstructured.Unstructured)
 	if !ok {
-		return fmt.Errorf("object to synchronize is expected to be Unstructured, but is %T", obj)
+		return nil, fmt.Errorf("object to synchronize is expected to be Unstructured, but is %T", obj)
 	}
 
 	if downstreamNamespace != "" {
 		if err := c.ensureDownstreamNamespaceExists(ctx, downstreamNamespace, upstreamObj); err != nil {
-			return err
+			return nil, err
 		}
 	} else {
 		// In cluser-wide resources we also need to check for possible collisions, as the resource could exist in the pcluster but now owned by this workspace.
 		// TODO(jmprusi): We should indicate the collision somehow (condition/annotation?) to avoid retrying the resource over and over.
 		if err := c.clusterWideCollisionCheck(ctx, gvr, upstreamObj); err != nil {
-			return err
+			return nil, err
 		}
 	}
 
-	// Make sure the DNS nameserver for the spec workspace is up and running
-	if err := c.dnsProcessor.Process(ctx, clusterName); err != nil {
-		logger.Error(err, "failed to check DNS nameserver is up and running (retrying)")
-		return err
+	deploymentGVR := schema.GroupVersionResource{
+		Group:    "apps",
+		Version:  "v1",
+		Resource: "deployments",
+	}
+	deploymentsSynced := false
+	if gvr == deploymentGVR {
+		deploymentsSynced = true
+	}
+
+	if _, exists := c.syncerInformers.InformerForResource(deploymentGVR); exists {
+		deploymentsSynced = true
```

Comment:
What's the meaning of checking for the informer existing, but not being synced?

Reply:
if deployments are in the GVRs mentioned in the `SyncedResources` status field of the `SyncedTarget`, there will be an informer started for deployments in `syncerInformers`.
In such a case, let's not wait for the first deployment instance to come, and start the DNS Deployment as soon as possible, so that it will be already started when the first deployment comes, and we don't need to wait before syncing the deployment.

Reply:
Sure - my point is that if you start doing this before the informer is synced you may have weird behavior

Reply:
Not sure understand why. I don't need the related informer here. In an informer has been started for deployments the means that deployments will very probably be synced by this Syncer in the current workspace, and will need an up and running CoreDNS POD dedicated to this workspace. So let's create it now.

Reply:
Hm - can we write this logic explicitly? Do we not have access to the configuration? Can we pass it down? Using an implicit check for a side-effect is hard to reason about.

Reply:
> Do we not have access to the configuration

What configuration are you speaking of ? The GVRs to be synced are dynamic and come from the current SyncTarget status.

Reply:
Let me add a comment.

Reply:
I removed the check BTW since having a `SyncTarget` that do not expect to sync `deployment`s is going to be very rare, and probably not worth a premature optimization.